### PR TITLE
Enable SE/HMC file access to repo (#1289918)

### PR DESCRIPTION
**Pull request by poncovka**
Provide access to files on the USB/DVD drive of the SE/HMC of System
z hardware with a boot option: inst.repo=hmc

Resolves: rhbz#1289918

Depends on: https://github.com/rhinstaller/pykickstart/pull/179
Depends on: https://github.com/rhinstaller/lorax/pull/251

### CODE REVIEW

File: dracut/anaconda-hmcroot [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+# anaconda-hmcroot:
+# Provide access to the root on the USB/DVD drive
+# of the SE/HMC of System z hardware.
```
Comment (jkonecny12) [inline dracut/anaconda-hmcroot]: I think the comment could be just one line.

File: dracut/anaconda-hmcroot [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+command -v getargbool >/dev/null || . /lib/dracut-lib.sh
+command -v config_get >/dev/null || . /lib/anaconda-lib.sh
```
Comment (jkonecny12) [inline dracut/anaconda-hmcroot]: If I understand this `command -v` correctly it is testing command availability if so why do you testing these commands if they are not used here anywhere?

File: dracut/anaconda-modprobe.sh [hunk 3]
```diff
@@ -30,6 +32,5 @@ MODULE_LIST+=" raid0 raid1 raid5 raid6 raid456 raid10 linear dm-mod dm-zero  \
               sha256 lrw xts "
 
 for m in $MODULE_LIST; do
-    modprobe $m &>/dev/null
+    modprobe $m && debug_msg "$m was loaded" || debug_msg "$m was NOT loaded"
```
Comment (jkonecny12) [inline dracut/anaconda-modprobe.sh]: Could this be an `if-else` instead of this please.

File: dracut/module-setup.sh [hunk 4]
```diff
@@ -14,6 +14,12 @@ depends() {
     return 0
 }
 
+installkernel() {
+    case "$(uname -m)" in
+        s390*) instmods hmcdrv ;;
+    esac
```
Comment (jkonecny12) [inline dracut/module-setup.sh]: Why to use `case` when it is only one value tested here?

File: dracut/module-setup.sh [hunk 5]
```diff
@@ -66,5 +72,14 @@ install() {
...
     done
+
+    # support for specific architectures
+    case "$(uname -m)" in
+        s390*)
+            inst "/usr/sbin/lshmc"
+            inst "/usr/bin/hmcdrvfs"
+            inst "$moddir/anaconda-hmcroot" "/sbin/anaconda-hmcroot"
+        ;;
+    esac
```
Comment (jkonecny12) [inline dracut/module-setup.sh]: The same as above. Why to use `case`?

File: pyanaconda/packaging/__init__.py [hunk 6]
```diff
@@ -1080,6 +1080,31 @@ def _setupInstallDevice(self, storage, checkmount):
...
+                log.debug("HMC is already mounted at %s.", DRACUT_REPODIR)
+                url = "file://" + DRACUT_REPODIR
+            else:
+                # Test the SE/HMC file access.
+                if iutil.execWithRedirect("/usr/sbin/lshmc", []):
+                    raise PayloadSetupError("The content of HMC media drive couldn't be accessed.")
+
+                # Test if a path is a mount point.
+                if os.path.ismount(INSTALL_TREE):
+                    log.debug("Don't mount the content of HMC media drive yet.")
```
Comment (jkonecny12) [inline pyanaconda/packaging/__init__.py]: I think there should be more something like `Test if HMC is a mount point path.`. The actual message is without context what is happening, you just know what is not happening.

File: pyanaconda/packaging/__init__.py [hunk 7]
```diff
@@ -1080,6 +1080,31 @@ def _setupInstallDevice(self, storage, checkmount):
...
+                else:
+                    # Make sure that the directories exists.
+                    iutil.mkdirChain(INSTALL_TREE)
+
+                    # Mount the device.
+                    if iutil.execWithRedirect("/usr/bin/hmcdrvfs", [INSTALL_TREE]):
+                        raise PayloadSetupError("The content of HMC media drive couldn't be mounted.")
+
+                # We are ready to use the HMC.
```
Comment (jkonecny12) [inline pyanaconda/packaging/__init__.py]: I think you can remove this comment. The debug message is pretty clear.

File: dracut/module-setup.sh (hunk 4 above)
Comment (poncovka) [reply dracut/module-setup.sh]: It is the same as in the function `depends` above. I think that the point is that we can more easily extend it with other cases.

File: dracut/anaconda-hmcroot (hunk 1 above)
Comment (poncovka) [reply dracut/anaconda-hmcroot]: Fixed.

File: dracut/anaconda-hmcroot (hunk 2 above)
Comment (poncovka) [reply dracut/anaconda-hmcroot]: Fixed, but you test the availability of a library, so it doesn't matter what command you test.

File: pyanaconda/packaging/__init__.py (hunk 6 above)
Comment (poncovka) [reply pyanaconda/packaging/__init__.py]: Fixed.

File: dracut/anaconda-modprobe.sh (hunk 3 above)
Comment (poncovka) [reply dracut/anaconda-modprobe.sh]: Fixed.

File: pyanaconda/packaging/__init__.py (hunk 7 above)
Comment (poncovka) [reply pyanaconda/packaging/__init__.py]: Fixed.

**Comment by poncovka** [comment]
Fixed the suggested changes.

File: dracut/anaconda-hmcroot (hunk 2 above)
Comment (jkonecny12) [reply dracut/anaconda-hmcroot]: It depends, you can for example rename the command or deprecate it, so there should be commands which you want to use to have valid feedback.

File: dracut/module-setup.sh (hunk 4 above)
Comment (jkonecny12) [reply dracut/module-setup.sh]: Ok.

File: pyanaconda/packaging/__init__.py [hunk 8]
```diff
@@ -1080,6 +1080,32 @@ def _setupInstallDevice(self, storage, checkmount):
             url = method.url
             mirrorlist = method.mirrorlist
             sslverify = not (method.noverifyssl or flags.noverifyssl)
+        elif method.method == "hmc":
+            # Check if /dev/hmcdrv is already mounted.
```
Comment (jkonecny12) [inline pyanaconda/packaging/__init__.py]: Could you please move this to its own method? This whole method is big as hell already and I don't think we want to have it even bigger.

**Comment by poncovka** [comment]
Jenkins, test this please

**Comment by poncovka** [comment]
Jenkins, test this please

**Comment by jkonecny12** [comment]
jenkins, test this please

**Pull request merged**